CRM/Utils/System/Standalone.php

```diff
@@ -569,13 +569,17 @@ public function permissionDenied() {
     // If not logged in, they need to.
     if (CRM_Core_Session::singleton()->get('ufID')) {
       // They are logged in; they're just not allowed this page.
+
+      // Check the (misconfiguration?) case that they're not allowed the civi homepage.
+      $item = CRM_Core_Invoke::getItem('civicrm/');
+      if (!CRM_Core_Permission::checkMenuItem($item)) {
+        throw new RuntimeException("User is allowed to login but does not have the requisite 'access CiviCRM' permission.");
+      }
       CRM_Core_Error::statusBounce(ts("Access denied"), CRM_Utils_System::url('civicrm'));
     }
     else {
       CRM_Utils_System::redirect('/civicrm/login?anonAccessDenied');
     }
```

Comment:
Aside: I think that the recent change in this area has caused some regressions in [E2E_Core_ErrorStatus]( https://test.civicrm.org/job/CiviCRM-E2E-Matrix/BKPROF=max,BLDTYPE=standalone-clean,CIVIVER=master,label=bknix-tmp/6487/testReport/(root)/E2E_Core_ErrorTest/testErrorStatus_with_data_set__backend_permission_/) -- i.e. if you send a web-service request without any credentials, then 403 makes sense.

I guess we probably need some heuristic to distinguish whether a 403 or redirect is more appropriate. Like... just spitballing...

* Based on path... `civicrm/ajax/*` => HTTP 403; other paths => HTTP Redirect
* Based on content-type... HTML => HTTP Redirect; other types => HTTP 403
* Based on user-agent... Mozilla => HTTP Redirect; other agents => HTTP 403
* Don't choose. Send HTTP 403 with an HTML document. Use `<meta>` refresh or a JS redirect.

Reply:
> If it's testing an ajax/server-to-server response then yes, I agree, it should not redirect to login, and should emit a 403.

Exactly

> But I don't get how the special frontend://civicrm/dev/fake-error map; It's not clear what is being tested?

In our discussion, we're making the distinction between scenarios where we do and don't want redirects, so I think you're looking closely. But the tests come out of a context where that cleverness hasn't existed AFAIK. On vanilla D7+D10, if anonymous requests an unauthorized resource, then they simply get HTTP 403. On vanilla WP, similar requests in `/wp-admin/` simply get HTTP 302.

When it comes to regular HTML pages hosted in CMSs... IMHO, the choice of 403 vs 302 isn't really our problem... like... those policies are liable to flipflop arbitrarily, and it doesn't really matter either way for us. Where it does become a concern is interop for AJAX/REST/web-hook/web-service. Ideally, tests would identify the approaches that give consistent behavior. (But if that's not possible, then the tests can at least show the typical behaviors - so that we can forewarn web-service consumers.)

> What about something like...

Yeah, using path `civicrm/ajax` feels like an "80% solution". Checking for GET sounds like a good idea -- it's probably another 80% solution. (*It's common+easy enough for REST and webhook callers to use POST.*) The combination is like 95%+. Much better than status-quo. For that last 5%, I'd probably lean toward a flag on the route XML (like `<web-service>true</web-service>` or `<env>web-service</env>`). (*I think there are additional reasons to have that flag...*)

I probably wouldn't bother with the `X-Civi-Auth` bit -- that's going too low level into auth mechanism, and I don't think it would improve the coverage enough to justify the extra complexity to the reader/tester/debugger. (*Ex: Doesn't help web-hooks.*)  The basic check for a logged-in user (*ufID*) is all we need from the auth system.